src/gtc/gtir.py

```diff
@@ -190,6 +193,53 @@ class VerticalLoop(LocNode):
     temporaries: List[FieldDecl]
     body: List[Stmt]
 
+    @root_validator(skip_on_failure=True)
+    def no_write_and_read_with_horizontal_offset(
+        cls, values: RootValidatorValuesType
+    ) -> RootValidatorValuesType:
+        """
+        In the same VerticalLoop a field must not be written and read with a horizontal offset.
+
+        Temporaries don't have this contraint. Backends are required to implement temporaries with block-private halos.
+        """
+        # TODO(havogt): either move to eve or will be removed in the attr-based eve if a List[Node] is represented as a CollectionNode
+        @utils.as_xiter
+        def _collection_iter_tree(
+            collection: List[Node],
+        ) -> Generator[TreeIterationItem, None, None]:
+            for elem in collection:
+                yield from elem.iter_tree()
+
+        def _writes(stmts: List[Stmt]) -> Set[str]:
+            result = set()
+            for left in _collection_iter_tree(stmts).if_isinstance(ParAssignStmt).getattr("left"):
+                result |= left.iter_tree().if_isinstance(FieldAccess).getattr("name").to_set()
+            return result
+
+        def _reads_with_offset(stmts: List[Stmt]) -> Set[str]:
+            return (
+                _collection_iter_tree(stmts)
+                .if_isinstance(FieldAccess)
+                .filter(
+                    lambda acc: acc.offset.i != 0 or acc.offset.j != 0
+                )  # writes always have zero offset
+                .getattr("name")
+                .to_set()
+            )
+
+        writes = _writes(values["body"])
+        reads_with_offset = _reads_with_offset(values["body"])
+
+        intersec = writes.intersection(reads_with_offset)
+        non_tmp_fields = set(
+            filter(lambda acc: acc not in [tmp.name for tmp in values["temporaries"]], intersec)
+        )
```

Comment:
```suggestion
        non_tmp_fields = {acc for acc in intersec if acc not in {tmp.name for tmp in values["temporaries"]}}
```
I find (set) comprehensions more readable than filter + conversion, but that’s a matter of preference I guess…